earth_enterprise/src/common/FileAccessor.h

```diff
@@ -0,0 +1,54 @@
+#include <string>
+#include <memory>
+#include <khFileUtils.h>
+
+class AbstractFileAccessor {
+public:
+  static std::unique_ptr<AbstractFileAccessor> getAccessor(const std::string &fname);
+  virtual bool isValid() = 0;
+  virtual void invalidate() = 0;
+  virtual int getFD() = 0;
+  virtual int Open(const std::string &fname, int flags, mode_t createMask = 0666) = 0;
+  virtual void Open(const char *fname, const char *mode) = 0;
+  virtual int FsyncAndClose() = 0;
+  virtual int Close() = 0;
+  virtual bool PreadAll(void* buffer, size_t size, off64_t offset) = 0;
+  virtual bool PwriteAll(const void* buffer, size_t size, off64_t offset) = 0;
+  virtual bool ReadStringFromFile(const std::string &filename, std::string &str, std::uint64_t limit = 0) = 0;
+  virtual bool Exists(const std::string &filename) = 0;
+  virtual int feof() = 0;
+  virtual void fgets(char *buf, int bufsize) = 0;
+  virtual void fprintf(const char * __restrict format, ...) = 0;
+};
+
+class POSIXFileAccessor: public AbstractFileAccessor {
+  int fileDescriptor;
```

Comment:
I don't love that there are two different ways to refer to a file. This is likely to cause trouble in the future if someone opens a file with a file descriptor and then calls a function that uses a file pointer. Is there any way we can combine them. Normally I'd suggest switching to all `FILE *`, but with the kh* functions taking file descriptors maybe it would be easier to convert the function that use a file pointer to use a file descriptor instead.

Reply:
On the flip side, I'm not sure if file descriptors support eof. Still, it would be best to avoid this duplication, or at least add a bunch of asserts that the user is calling the right function (so at the beginning of a function that uses a file descriptor, assert that the file descriptor is valid. Similar for file pointers).

Reply:
I was thinking about this as well. One thought would be to only use the file descriptor by default. But if a client calls one of the methods like `fgets` or `fprintf` that need a file pointer, you could call `fdopen` internally to get a `FILE*` from the file descriptor. In that case, you could store the `FILE*` in a member variable that normally doesn't get used. And closing one will close the other (or at least closing the `FILE*` will close the file descriptor as well).

Reply:
Also, in that implementation, you may be able to get rid of the second `Open` method: `virtual void Open(const char *fname, const char *mode) = 0;`

Reply:
As of now, fileDescriptor is just an int. Are you suggesting making it it's own class?

Reply:
No, I'm just talking about making POSIXFileAccessor a bit smarter. 

1. Get rid of the second `Open` method and just add an optional parameter, `mode`, to the first one.
2. Always initialize the `fp` member to `nullptr` in the constructor
3. In the member methods that use the `fp` member, just make sure it's initialized as follows.
`void POSIXFileAccessor::fgets(char *buf, int bufsize) {
 
  VerifyFileStream();
 ::fgets(buf, bufsize, fp);
}`

With `VerifyFileStream` defined something like
`inline void POSIXFileAccessor::VerifyFileStream() {
	if (!fp)
		fp = fdopen(fileDescriptor, mode);
}`

I might also modify the `Close` method just slightly to check the `fp` first. If you call `fclose`, I know it closes the file descriptor as well. I don't know if the reverse is true for `close`.

`int POSIXFileAccessor::Close() {
  int result = 0;
  if (fp) {
    result = ::fclose(fp);
  }
  else {
    result = khClose(fileDescriptor);
  }
  this->invalidate();
  return result;
}`

Reply:
I _think_ that should work, but I haven't tried to build it with those changes.